accumulator_updater/programs/accumulator_updater/src/instructions/emit_inputs.rs

```diff
@@ -0,0 +1,165 @@
+use {
+    crate::{
+        accumulator_acc_seeds,
+        accumulator_acc_seeds_with_bump,
+        state::*,
+        AccumulatorUpdaterError,
+    },
+    anchor_lang::{
+        prelude::*,
+        system_program::{
+            self,
+            CreateAccount,
+        },
+    },
+};
+
+
+pub fn emit_inputs<'info>(
+    ctx: Context<'_, '_, '_, 'info, EmitInputs<'info>>,
+    base_account: Pubkey,
+    data: Vec<Vec<u8>>,
+    account_type: u32,
+    account_schemas: Vec<u8>,
+) -> Result<()> {
+    let cpi_caller = ctx.accounts.whitelist_verifier.is_allowed()?;
+    let accts = ctx.remaining_accounts;
+    require_eq!(accts.len(), data.len());
+    require_eq!(data.len(), account_schemas.len());
+    let mut zip = data.into_iter().zip(account_schemas.into_iter());
+
+    let rent = Rent::get()?;
+    let (mut initialized, mut updated) = (vec![], vec![]);
+    for ai in accts {
+        let (account_data, account_schema) = zip.next().unwrap();
+
+        if is_uninitialized_account(ai) {
+            let seeds = accumulator_acc_seeds!(cpi_caller, base_account, account_schema);
+            let (pda, bump) = Pubkey::find_program_address(seeds, &crate::ID);
+            require_keys_eq!(ai.key(), pda);
+
+            //TODO: Update this with serialization logic
+            let accumulator_size = 8 + AccumulatorInput::get_initial_size(&account_data);
+            let accumulator_input = AccumulatorInput::new(
+                AccumulatorHeader::new(
+                    bump,
+                    1, //from CPI caller?
+                    account_type,
+                    account_schema,
+                ),
+                account_data,
+            );
+            EmitInputs::init_accumulator_input(
+                ai,
+                accumulator_input,
+                accumulator_size,
+                &ctx.accounts.payer,
+                &[accumulator_acc_seeds_with_bump!(
```

Comment:
i'm a little confused about the use of this function w/ the bump vs. the one w/o the bump. It seems like you want there to be a single PDA for each key, right? if so, isn't the bump deterministically generated, so there shouldn't be a reason to save it? (is this an efficiency improvement? if so, does it not introduce security holes because there are now multiple PDAs that can pass the checks for the same key?)

Reply:
there's the "canonical bump" which is the first bump counting down from 255 that provides a valid PDA address (off the ed25519 curve). This is the bump that's returned with `find_program_address`. But there could technically be multiple bumps that create valid PDA addresses. 

There's 2 ways to verify a PDA address:
1. assume you're always using the canonical bump and check that `Pubkey::find_program_address(&[seeds]).0 == account.key()`
2. store the bump in the account and check`account.key() == Pubkey::create_program_address(&[seeds, &[account.bump]])` 

2 is more efficient (marginally) since 1 could theoretically have to do a guess-and-check multiple times to find the first valid seed. 

Regarding your concern about security holes - since only a program can create PDAs it owns, it shouldn't be possible to have multiple PDAs using the same seeds but different bumps. The accumulator-updater program will be the only program to create these PDAs and it's using the canonical bumps when the PDAs are first initialized.

Having 2 macros is a bit confusing so i'll remove the non-bump one since that's only ever used once